Document how to migrate from Ingress to Gateway API (#25599)
cilium-team/cilium: Documentation/network/servicemesh/ingress-to-gateway/http-migration.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _gs_gateway_http_migration:

**********************
HTTP Migration Example
**********************

> In this example, we will show how to take an existing Ingress configuration and migrate it to the equivalent Gateway API resource.

Reviewer: Address the user as "you", here and throughout the document.
```suggestion
This example shows you how to migrate an existing Ingress configuration to the equivalent Gateway API resource.
```

We will use the Cilium :ref:`gs_ingress_http` as the starting Ingress configuration. 

> The same approach will apply to other controllers, albeit each Ingress controller configuration varies.

Reviewer: Address the user as "you", use present tense, use clear, plain language
```suggestion
The Cilium :ref:`gs_ingress_http` serves as the starting Ingress configuration. 
The same approach applies to other controllers, though each Ingress controller configuration varies.
```

The example Ingress configuration routes traffic to backend services from the
``bookinfo`` demo microservices app from the Istio project.

Review Ingress Configuration
============================

> You'll find the example Ingress definition in ``basic-ingress.yaml``.

Reviewer: Use present tense
```suggestion
You can find the example Ingress definition in ``basic-ingress.yaml``.
```

.. literalinclude:: ../../../../examples/kubernetes/servicemesh/basic-ingress.yaml

This example listens for traffic on port 80, routes requests for the path ``/details`` to the ``details`` service,
and ``/`` to the ``productpage`` service.

Create Equivalent Gateway Configuration
=======================================

> To create the equivalent Gateway configuration, you must consider the following:

Reviewer: Reserve "must" for actions a user _must_ take to succeed, rather than stressing the importance of general considerations.
```suggestion
To create the equivalent Gateway configuration, consider the following:
```

- Entry Point

> The entry point is a combination of an IP address and port through which the data plane is accessible to external clients.

Reviewer: Use active voice
```suggestion
The entry point is a combination of an IP address and port through which external clients access the data plane.
```

.. tabs::

    .. group-tab:: Ingress
        
        Every Ingress resource has two implicit entry points -- one for HTTP and the other for HTTPS traffic. 

>         An Ingress controller provides those entry points. Typically, they will either be shared by all Ingress resources, or every Ingress resource will get dedicated entry points.

Reviewer: Clarity, use present tense, specific is better than general
```suggestion
        An Ingress controller provides the entry points. Typically, entry points are either shared by all Ingress resources, or every Ingress resource has dedicated entry points.
```

        .. code-block:: shell-session

            apiVersion: networking.k8s.io/v1
            kind: Ingress
            spec:
              ingressClassName: cilium

    .. group-tab:: Gateway API

>         In the Gateway API, entry points must be explicitly defined in a Gateway resource. 

Reviewer: Here's an example of where "must" is appropriate. 👍🏻

>         For example, if you want the data plane to handle HTTP traffic on port 80, you need to define a listener for that traffic. 

Reviewer: Use clear, plain language; here's an example of where "must" is appropriate
```suggestion
        For example, for the data plane to handle HTTP traffic on port 80, you must define a listener for that traffic. 
```

        Typically, a Gateway implementation provides a dedicated data plane for each Gateway resource.

        .. code-block:: shell-session

            apiVersion: gateway.networking.k8s.io/v1beta1
            kind: Gateway
            metadata:

>               name: cilium

Reviewer 2: I would maybe not reccomend against naming this `cilium` this will become less clear what a gateway is when trying it out

Author: Changed the gateway names from `cilium` to `cilium-gateway`.

            spec:
              gatewayClassName: cilium
              listeners:
              - name: http
                port: 80
                protocol: HTTP

- Routing Rules

> When using Ingress or Gateway API, routing rules need to be defined to attach applications to those entry points.

Reviewer: clarity
```suggestion
When using Ingress or Gateway API, routing rules must be defined to attach applications to those entry points.
```

.. tabs::

    .. group-tab:: Ingress
        

>         The path-based routing rules are configured in the Ingress resource map directly to the routing rules of the HTTPRoute.

Reviewer: > routing rules are configured in the Ingress resource map directly to the routing rules

Reword this to make the action/path of rules clearer.

>         In the Ingress, each hostname has separate routing rules, 

Reviewer: ```suggestion
        In the Ingress, each hostname has separate routing rules:
```

        .. code-block:: shell-session

            apiVersion: networking.k8s.io/v1
            kind: Ingress
            [...]
            rules:
            - http:
                paths:
                - backend:
                    service:
                      name: details
                      port:
                        number: 9080
                  path: /details
                  pathType: Prefix
                - backend:
                    service:
                      name: productpage
                      port:
                        number: 9080
                  path: /
                  pathType: Prefix

    .. group-tab:: Gateway API

        The routing rules are configured in the HTTPRoute.

        .. code-block:: shell-session

            ---
            apiVersion: gateway.networking.k8s.io/v1beta1
            kind: HTTPRoute
            spec:
              parentRefs:
              - name: cilium

>             rules:

Reviewer 2: This feels stupid but how about we turn these around, they would read more like a `from -> to` in my head

```yaml
            rules:
            - matches:
              - path:
                  type: PathPrefix
                  value: /
              backendRefs:
              - name: productpage
                port: 9080
```

Author: Makes perfect sense. Changed these around.

            - backendRefs:
              - name: productpage
                port: 9080
              matches:
              - path:
                  type: PathPrefix
                  value: /
            - backendRefs:
              - name: details
                port: 9080
              matches:
              - path:
                  type: PathPrefix
                  value: /details

- Selecting Data Plane to Attach to:

> Both Ingress and Gateway API resources need to be explicitly attached to a Dataplane.  

Reviewer: must
```suggestion
Both Ingress and Gateway API resources must be explicitly attached to a Dataplane.  
```

.. tabs::

    .. group-tab:: Ingress

>         An Ingress resource must specify a class to select which Ingress controller to use. 

Reviewer: ```suggestion
        An Ingress resource must specify a class that selects which Ingress controller to use. 
```

        .. code-block:: shell-session

            apiVersion: networking.k8s.io/v1
            kind: Ingress
            spec:
              ingressClassName: cilium

    .. group-tab:: Gateway API

>         A Gateway resource must also specify a class: in our case, it will always be the ``cilium`` class. 

Reviewer: Address the user as "you", use present tense
```suggestion
        A Gateway resource must also specify a class: in this example, it is always the ``cilium`` class. 
```

        An HTTPRoute must specify which Gateway (or Gateways) to attach to via a ``parentRef``.

        .. code-block:: shell-session

            apiVersion: gateway.networking.k8s.io/v1beta1
            kind: Gateway
            metadata:
              name: cilium
              namespace: default
            spec:
              gatewayClassName: cilium
            [...]
            ---
            apiVersion: gateway.networking.k8s.io/v1beta1
            kind: HTTPRoute
            spec:
              parentRefs:
              - name: cilium

Review Equivalent Gateway Configuration
=======================================

> You'll find the equivalent final Gateway and HTTPRoute definition in ``http-migration.yaml``.

Reviewer: Use present tense
```suggestion
You can find the equivalent final Gateway and HTTPRoute definition in ``http-migration.yaml``.
```

.. literalinclude:: ../../../../examples/kubernetes/gateway/http-migration.yaml

> The above example creates a Gateway named ``cilium`` that listens on port 80 for HTTP traffic.

Reviewer: s/above/preceding/g
```suggestion
The preceding example creates a Gateway named ``cilium`` that listens on port 80 for HTTP traffic.
```
